### PR TITLE
feat(FileAction): add file action support

### PR DESCRIPTION
**Pull request by skjnldsv**
RFC: https://github.com/nextcloud/server/issues/21835

``` ts
interface FileActionData {
	id: string
	displayName: (files: Node[], view) => string
	iconSvgInline: (files: Node[], view) => string
	enabled?: (files: Node[], view) => boolean
	exec: (file: Node, view) => Promise<boolean>,
	execBatch?: (files: Node[], view) => Promise<boolean[]>
	order?: number,
	default?: boolean,
	inline?: (file: Node, view) => boolean,
	renderInline?: (file: Node, view) => HTMLElement,
}
```

### CODE REVIEW

File: lib/fileAction.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	/** Svg as inline string. <svg><path fill="..." /></svg> */
+	iconSvgInline: (files: Node[]) => string
+	/** Condition wether this action is shown or not */
+	enabled?: (files: Node[], view) => boolean
+	/**
+	 * Function executed on single file action
+	 * @returns true if the action was executed, false otherwise
+	 * @throws Error if the action failed
+	 */
+	exec: (file: Node) => boolean,
```
Comment (ChristophWurst) [inline lib/fileAction.ts:40]: would it make sense to make these async and awaitable?

Comment (skjnldsv) [reply lib/fileAction.ts]: Done
Also edited first post

**Comment by skjnldsv** [comment]
The default is the last thing I'm not sure of.
This is very different than what we do right now

Same for the registering, the legacy way is registering per mimetype
And registering as default is done as a secondary step

**Comment by skjnldsv** [comment]
After a bit of thoughts, and seeing that there are very few number of actions I don't think we would really gain by splitting per mime.
Let's just have one big list and let apps register their own actions conditions

Instead of
```js
['image/jpeg', 'image/png', ...].forEach(mime => registerAction(action, mime))
```

It's easier and cleaner to do
```js
registerAction(action)
```

When opening the actions menu, then it would make it easy to just do something like this (which we would have to do anyway even if we group per mimes)
```vue
<template>
	<Actions>
		<ActionButton v-for="action in enabled" ... />
	</Actions>
</template>
<script>
const actions = getActions()
const enabled = actions.filter(action => action.enabled(this.file, this.currentView))
</script>
```

@juliushaertl any opinion?

**Comment by juliushaertl** [comment]
I think the enabled option makes sense. This way developers have one place for implementing their logic. 

I'd be a bit hesitant about the boolean value for default. From the API usage perspective the difference to the current way should not matter, but both the old and new approach lack way to properly handle multiple actions registering as a default. The current approach is that the last registered wins. I'm not sure about a better approach here tough. We could also allow passing a callback here so that apps can make this dependent on file attributes (like the disabled download), but this could also be skipped I guess if we consider viewer as the preferred default opener. One use case for this would be https://github.com/nextcloud/viewer/issues/1440 where we would want to open pdf files with files_pdfviewer by default, except when secure view of collabora should be used.

**Comment by juliushaertl** [comment]
The only downside this has is that app developers need to align how to handle those scenarios. But as long as we don’t decide to offer preferences for default apps or actions this seems like the most reasonable option.

**Comment by skjnldsv** [comment]
> But as long as we don’t decide to offer preferences for default apps or actions this seems like the most reasonable option.

Could you elaborate? :)

**Comment by juliushaertl** [comment]
Mostly about https://github.com/nextcloud/server/issues/8105 (and related feature requests we frequently get for Collabora) where there could be more than one app that provides a file handler and either the admin or users would prefer having the choice about the default.

I don't think we need to have this as part of the initial API design, just something to keep in mind that this might be a topic in the future. Especially with a growing app ecosystem.

**Comment by juliushaertl** [comment]
Thinking about this again, I'd also say that the boolean value is good and the preference could still be handled with that. Maybe we just need to clarify that "default" only means that the registering app would like to be the default, but there are no guarantees about that due to the first-wins approach at the moment.

**Comment by skjnldsv** [comment]
> I don't think we need to have this as part of the initial API design, just something to keep in mind that this might be a topic in the future. Especially with a growing app ecosystem.

Could be easily done afterwards if we have multiple defaults.
The first one would be used as default click, while the others are put in the menu :+1:

**Comment by skjnldsv** [comment]
Great!

You should know this API is already being used on the trashbin branch and is working pretty great (with a few adjustments that i'll push later) :+1: :rocket:

**Comment by skjnldsv** [comment]
Done everyone!

**Review by artonge — APPROVED** [review]
Looks good :+1:

**Pull request merged**